ESMF/Apps/esm.F90

```diff
@@ -0,0 +1,160 @@
+
+    implicit none
+
+    private
+
+    public SetServices
+
+!-----------------------------------------------------------------------------
+contains
+!-----------------------------------------------------------------------------
+
+    subroutine SetServices(driver, rc)
+        type(ESMF_GridComp)  :: driver
+        integer, intent(out) :: rc
+
+        rc = ESMF_SUCCESS
+
+        print*, "ESM start SetServices"
+
+        print*, "ESM register Generic SetServices"
+        ! NUOPC_Driver registers the generic methods
+        call NUOPC_CompDerive(driver, driver_routine_SS, rc=rc)
+        VERIFY_NUOPC_(rc)
+
+        ! attach specializing method(s)
+        print*, "ESM attach SetModelServices"
+        call NUOPC_CompSpecialize(driver, &
+                specLabel=driver_label_SetModelServices, &
+                specRoutine=SetModelServices, rc=rc)
+        VERIFY_NUOPC_(rc)
+
+        ! set driver verbosity
+        print*, "ESM set attribute"
+        call NUOPC_CompAttributeSet(driver, name="Verbosity", value="low", rc=rc)
+        VERIFY_NUOPC_(rc)
+
+        print*, "ESM finish SetServices"
+    end subroutine
+
+!-----------------------------------------------------------------------------
+
+    subroutine SetModelServices(driver, rc)
+        type(ESMF_GridComp)  :: driver
+        integer, intent(out) :: rc
+
+        ! local variables
+        type(ESMF_Grid)               :: grid
+        type(ESMF_Field)              :: field
+        type(ESMF_Time)               :: startTime
+        type(ESMF_Time)               :: stopTime
+        type(ESMF_TimeInterval)       :: timeStep
+        type(ESMF_Clock)              :: internalClock
+        type(ESMF_GridComp)           :: child, test
+        type(ESMF_CplComp)            :: connector
+
+        rc = ESMF_SUCCESS
+
+        print*, "ESM start SetModelServices"
+
+        ! SetServices for ATM
+        print*, "ESM add ATM comp"
+        call NUOPC_DriverAddComp(driver, "ATM", atmSS, comp=child, rc=rc)
+        VERIFY_NUOPC_(rc)
+        call NUOPC_CompAttributeSet(child, name="Verbosity", value="low", rc=rc)
+        VERIFY_NUOPC_(rc)
+
+        call NUOPC_DriverAddComp(driver, "test", NUOPC_CAP_SS, comp=child, rc=rc)
+        VERIFY_NUOPC_(rc)
+
+        ! SetServices for OCN
+        print*, "ESM add OCN comp"
+        call NUOPC_DriverAddComp(driver, "OCN", ocnSS, comp=child, rc=rc)
+        VERIFY_NUOPC_(rc)
+        call NUOPC_CompAttributeSet(child, name="Verbosity", value="low", rc=rc)
+        VERIFY_NUOPC_(rc)
+
+        ! Disabling the following macro, e.g. renaming to WITHCONNECTORS_disable,
+        ! will result in a driver that does not call connectors between the model
+        ! components. This mode can be used if all model components are driven 
+        ! as independent models. However, even for independent models the
+        ! connectors can be set here, but will turn into no-ops.
+#define WITHCONNECTORS
+#ifdef WITHCONNECTORS
```

Comment:
We generally try to avoid CPP conditionals in GEOS software.  Are these essential?

Reply:
Will remove